nixos/modules/services/web-apps/jitsi-meet.nix

```diff
@@ -0,0 +1,503 @@
+          storage = "memory"
+          -- muc_room_cache_size = 1000
+          admins = { "focus@auth.${cfg.hostName}", "jvb@auth.${cfg.hostName}" }
+
+        Component "focus.${cfg.hostName}"
+          component_secret = os.getenv("JICOFO_COMPONENT_SECRET")
+      '';
+      virtualHosts.${cfg.hostName} = {
+        enabled = true;
+        domain = cfg.hostName;
+        extraConfig = ''
+          authentication = "anonymous"
+          c2s_require_encryption = false
+          admins = { "focus@auth.${cfg.hostName}" }
+        '';
+      };
+      virtualHosts."auth.${cfg.hostName}" = {
+        enabled = true;
+        domain = "auth.${cfg.hostName}";
+        extraConfig = ''
+          authentication = "internal_plain"
+        '';
+      };
+    };
+    systemd.services.prosody.serviceConfig = mkIf cfg.prosody.enable {
+      EnvironmentFile = [ "/var/lib/jitsi-meet/secrets" ];
+      SupplementaryGroups = [ "jitsi-meet" ];
+    };
+
+    users.groups.jitsi-meet = {};
+    systemd.tmpfiles.rules = [
+      "d '/var/lib/jitsi-meet' 0750 root jitsi-meet - -"
+    ];
+
+    systemd.services.jitsi-meet-init-secrets = {
+      wantedBy = [ "multi-user.target" ];
+      before = [ "jicofo.service" "jitsi-videobridge2.service" ] ++ (optional cfg.prosody.enable "prosody.service");
+      serviceConfig = {
+        Type = "oneshot";
+      };
+
+      script = let
+        secrets = [ "VIDEOBRIDGE_SECRET" "JICOFO_COMPONENT_SECRET" "JICOFO_USER_SECRET" ];
+      in
+      ''
+        if [ ! -f /var/lib/jitsi-meet/secrets ]; then
+          ${concatMapStringsSep "\n" (var: ''
+            echo "${var}=$(tr -dc a-zA-Z0-9 </dev/urandom | head -c 64)" >> /var/lib/jitsi-meet/secrets
+          '') secrets}
+
+          chown root:jitsi-meet /var/lib/jitsi-meet/secrets
+          chmod 640 /var/lib/jitsi-meet/secrets
+        fi
+
+      ''
+      + optionalString cfg.prosody.enable ''
+        source /var/lib/jitsi-meet/secrets
+        ${config.services.prosody.package}/bin/prosodyctl register focus auth.${cfg.hostName} "$JICOFO_USER_SECRET"
+        ${config.services.prosody.package}/bin/prosodyctl register jvb auth.${cfg.hostName} "$VIDEOBRIDGE_SECRET"
+
+        cd ${config.services.prosody.dataDir}
+
+        # regenerate the certificates after 1 month - they are valid for a year, that leaves 11 months for service restart
+        find ${cfg.hostName}.crt -mtime +30 -delete || true
+
+        # generate self-signed certificates
+        if [ ! -f ${cfg.hostName}.crt ]; then
+          export PATH=$PATH:${pkgs.openssl.bin}/bin
+          rm -f {,auth.}${cfg.hostName}.{crt,key,cnf}
+          # not sure this one is even used but it avoids some errors in logs
+          echo | ${config.services.prosody.package}/bin/prosodyctl cert generate ${cfg.hostName}
+          # used for jicofo<->prosody user connection
+          echo | ${config.services.prosody.package}/bin/prosodyctl cert generate auth.${cfg.hostName}
+        fi
```

Comment:
As already expressed, we have a somewhat usable ACME module. I'd prefer very much to rely on it to generate certificates, with proper timers and reload logic, instead of creating them inside a `jitsi-meet-init-secrets` oneshot unit, which will only happen when the box is rebooted or the service restarted.

Reply:
I would strongly recommend _against_ using ACME unless you have a pressing need to. For a single-host setup, setting up the nginx host and requesting certificates for `auth.${hostName}` is totally unnecessary.

For people who really need ACME, it could be provided, but I would want to keep something like this as the default.

Reply:
What's the problem with using the ACME infrastructure to request these certificates?
We default to use ACME certificates, if no custom ones are provided.
I don't see why we shouldn't apply the same default here:
They're free, a nixpkgs-wide default, renewals are handled properly via timers, and we don't need to mess manually with trust stores.

Reply:
They are unnecessary, they require a web-server to respond on that particular hostname, and they count against your Let's Encrypt quotas.

I don't really see why you'd enable them unless you explicitly need them. Enabling them by default just means you're requesting public certificates from a public service to protect communication from localhost to localhost. I absolutely wouldn't want that behaviour without explicitly opting in.

Reply:
Ok, there seem to be some valid concerns on both sides. I have some technical questions to understand our options: 

1. How would using ACME here look like? If you use ACME for your nginx certificate, could the same certificate be used for prosody? Or do we need another one?
2. If we cannot reuse the certificate, would you need to expose a bogus webserver on that hostName for the ACME process? Or would you skip that in that case? Then we would need a fallback.
3. If we reuse the certificate (see 1), does it still have any impact on our ACME quota?
4. Do we have alternative infastructure in nixpkgs to generate bogus certificates that we could use here?

Reply:
1.  There should be a very similar example in https://github.com/NixOS/nixpkgs/pull/86067#issuecomment-619586870, but that PR is missing some documentation. We should be able to reuse certificates, as long as both prosody and nginx are in a shared group (IIRC)
2.  Currently the acme module supports verification via a running webserver serving the challenge files (with support built into the `nginx` module), as well as DNS-01 verification. If people don't want to use DNS-01, adding a small vhost for the other domains that redirects to the main jitsi instance shouldn't be too complicated
3.  wrt. quotas: we currently create a new user account per key in the `security.acme` attrset. Simply adding all jitsi-related domains to `extraDomains` shouldn't add more to the quota, and we need to create a certificate for the web frontend anyways.
4.  The acme module creates a bogus certificate until real certificates have been provided. I'm not aware of any nix-based nixpkgs-wide tooling to bootstrap a CA - I consider this a bad idea, as the private keys of the CA end up world-readable in the nix-store.

Reply:
Maybe it is acceptable for @ArdaXi to first support the use case where Jitsi is made public anyways and we reuse the certificate for prosody?

Later on, if needed, we could add config options to get the certificates from somewhere else...

Reply:
@kolloch I'd be fine with reusing the certificate, which would give you a certificate for `${hostName}`. That would still leave you with the `auth.${hostName}` certificate. None of the Prosody hostnames are public-facing, but `auth.${hostName}` also is not an nginx vHost and also does not have to be a public DNS entry at _all_.

Setting up the re-use for just `${hostName}` means you're using ACME for one hostname and generating and managing a self-signed cert for another hostname.

Using ACME for both requires all users to have a DNS entry for `auth.${hostName}` and a virtualHost entry in nginx that's just an ACME shim, all for a certificate that no server the user doesn't control will ever see, and in most cases a certificate that no other server than the one it's generated on will ever see.

I still don't see the point, to be honest. I think it's a bad idea to have any ACME-related functionality as a default because it would require the certificate's key to be readable by both nginx and prosody, and it would require the administrator to set up an additional DNS entry, both for no benefit other than "it's slightly cleaner on the implementation side."

If it's available as an option, sure I guess, but I would never want to use it.